File: lib/routing.lua
--- Canvas-to-device routing engine.
--
-- Builds a pre-computed routing table that maps each physical LED to the
-- canvas cells it overlaps, weighted by fractional area coverage.
-- On each canvas frame, the table is walked and colours are mixed with
-- area-weighted averaging.
--
-- Usage from init.lua:
--   local routing = require("lib.routing")
--   local table   = routing.build(placements, GRID_W, GRID_H)
--   local colors  = routing.route(canvas_colors, table)

local M = {}

local floor = math.floor
local ceil  = math.ceil
local min   = math.min
local max   = math.max
local sqrt  = math.sqrt
local abs   = math.abs
local sin   = math.sin
local cos   = math.cos
local rad   = math.rad
local type  = type
local tonumber = tonumber
local EPS   = 1e-9

--- Compute fractional overlap area between two axis-aligned rectangles.
-- Returns 0 when the rectangles do not overlap.
local function overlap_area(ax1, ay1, ax2, ay2, bx1, by1, bx2, by2)
    local ox = max(0, min(ax2, bx2) - max(ax1, bx1))
    local oy = max(0, min(ay2, by2) - max(ay1, by1))
    return ox * oy
end

--- Rotate a point (px, py) around center (cx, cy) by angle_deg degrees (clockwise).
local function rotate_point(px, py, cx, cy, cos_a, sin_a)
    local dx = px - cx
    local dy = py - cy
    return cx + dx * cos_a - dy * sin_a,
           cy + dx * sin_a + dy * cos_a
end

--- Compute the axis-aligned bounding box of a rotated rectangle.
-- Takes the four corners of an unrotated rect (x1,y1)-(x2,y2), rotates
-- them around (cx,cy) and returns the enclosing AABB.
local function rotated_rect_aabb(x1, y1, x2, y2, cx, cy, cos_a, sin_a)
    local ax, ay = rotate_point(x1, y1, cx, cy, cos_a, sin_a)
    local bx, by = rotate_point(x2, y1, cx, cy, cos_a, sin_a)
    local ccx, ccy = rotate_point(x2, y2, cx, cy, cos_a, sin_a)
    local dx, dy = rotate_point(x1, y2, cx, cy, cos_a, sin_a)

    local min_x = min(ax, bx, ccx, dx)
    local min_y = min(ay, by, ccy, dy)
    local max_x = max(ax, bx, ccx, dx)
    local max_y = max(ay, by, ccy, dy)

    return min_x, min_y, max_x, max_y
end

local function polygon_area(points)
    local count = #points
    if count < 3 then
        return 0
    end

    local area = 0
    local prev = points[count]
    for i = 1, count do
        local curr = points[i]
        area = area + (prev.x * curr.y) - (curr.x * prev.y)
        prev = curr
    end

    return abs(area) * 0.5
end

local function clip_polygon_left(points, left)
    local result = {}
    local prev = points[#points]
    if not prev then
        return result
    end

    local prev_inside = prev.x >= left - EPS
    for i = 1, #points do
        local curr = points[i]
        local curr_inside = curr.x >= left - EPS

        if prev_inside ~= curr_inside then
            local dx = curr.x - prev.x
            if abs(dx) > EPS then
                local t = (left - prev.x) / dx
                result[#result + 1] = {
                    x = left,
                    y = prev.y + (curr.y - prev.y) * t,
                }
            end
        end

        if curr_inside then
            result[#result + 1] = { x = curr.x, y = curr.y }
        end

        prev = curr
        prev_inside = curr_inside
    end

    return result
end

local function clip_polygon_right(points, right)
    local result = {}
    local prev = points[#points]
    if not prev then
        return result
    end

    local prev_inside = prev.x <= right + EPS
    for i = 1, #points do
        local curr = points[i]
        local curr_inside = curr.x <= right + EPS

        if prev_inside ~= curr_inside then
            local dx = curr.x - prev.x
            if abs(dx) > EPS then
                local t = (right - prev.x) / dx
                result[#result + 1] = {
                    x = right,
                    y = prev.y + (curr.y - prev.y) * t,
                }
            end
        end

        if curr_inside then
            result[#result + 1] = { x = curr.x, y = curr.y }
        end

        prev = curr
        prev_inside = curr_inside
    end

    return result
end

local function clip_polygon_top(points, top)
    local result = {}
    local prev = points[#points]
    if not prev then
        return result
    end

    local prev_inside = prev.y >= top - EPS
    for i = 1, #points do
        local curr = points[i]
        local curr_inside = curr.y >= top - EPS

        if prev_inside ~= curr_inside then
            local dy = curr.y - prev.y
            if abs(dy) > EPS then
                local t = (top - prev.y) / dy
                result[#result + 1] = {
                    x = prev.x + (curr.x - prev.x) * t,
                    y = top,
                }
            end
        end

        if curr_inside then
            result[#result + 1] = { x = curr.x, y = curr.y }
        end

        prev = curr
        prev_inside = curr_inside
    end

    return result
end

local function clip_polygon_bottom(points, bottom)
    local result = {}
    local prev = points[#points]
    if not prev then
        return result
    end

    local prev_inside = prev.y <= bottom + EPS
    for i = 1, #points do
        local curr = points[i]
        local curr_inside = curr.y <= bottom + EPS

        if prev_inside ~= curr_inside then
            local dy = curr.y - prev.y
            if abs(dy) > EPS then
                local t = (bottom - prev.y) / dy
                result[#result + 1] = {
                    x = prev.x + (curr.x - prev.x) * t,
                    y = bottom,
                }
            end
        end

        if curr_inside then
            result[#result + 1] = { x = curr.x, y = curr.y }
        end

        prev = curr
        prev_inside = curr_inside
    end

    return result
end

local function polygon_rect_overlap_area(points, x1, y1, x2, y2)
    local clipped = clip_polygon_left(points, x1)
    if #clipped == 0 then return 0 end
    clipped = clip_polygon_right(clipped, x2)
    if #clipped == 0 then return 0 end
    clipped = clip_polygon_top(clipped, y1)
    if #clipped == 0 then return 0 end
    clipped = clip_polygon_bottom(clipped, y2)
    if #clipped == 0 then return 0 end
    return polygon_area(clipped)
end

--- Determine the LED grid dimensions (cols, rows) for a placed device.
--
-- If the device has a matrix with explicit width/height we use those.
-- Otherwise we derive a nearly-square auto-grid from `leds_count`:
--   cols = ceil(sqrt(N)), rows = ceil(N / cols)
local function led_grid_dims(p)
    if p.matrix and p.matrix.width and p.matrix.width > 0
       and p.matrix.height and p.matrix.height > 0 then
        return p.matrix.width, p.matrix.height
    end
    local n = p.ledsCount or 0
    if n <= 0 then return 1, 1 end
    return n, 1
end

--- Build LED world rectangles indexed by LED id (0-based + 1 for Lua array).
-- Uses matrix.map when available so sparse/non-linear matrix layouts keep their
-- actual LED positions instead of assuming row-major LED ordering.
--
-- When rotation ~= 0, each LED rect is stored as a rotated quadrilateral plus
-- an axis-aligned bounding box used to coarse-filter candidate canvas cells.
local function build_led_rects(p, cols, rows, total_leds)
    local dev_w = (type(p.width) == "number" and p.width > 0) and p.width or cols
    local dev_h = (type(p.height) == "number" and p.height > 0) and p.height or rows
    local origin_x = (type(p.x) == "number") and p.x or 0
    local origin_y = (type(p.y) == "number") and p.y or 0
    local rotation = tonumber(p.rotation) or 0
    rotation = rotation % 360
    if rotation < 0 then
        rotation = rotation + 360
    end

    local led_w = dev_w / cols
    local led_h = dev_h / rows

    -- Precompute rotation if needed.
    local has_rotation = abs(rotation) > EPS
    local cos_a, sin_a, center_x, center_y
    if has_rotation then
        local angle_rad = rad(rotation)
        cos_a = cos(angle_rad)
        sin_a = sin(angle_rad)
        center_x = origin_x + dev_w / 2
        center_y = origin_y + dev_h / 2
    end

    local rects = {}
    local matrix = p.matrix
    local map = matrix and matrix.map

    local function make_led_geometry(x1, y1, x2, y2)
        if not has_rotation then
            return {
                x1 = x1,
                y1 = y1,
                x2 = x2,
                y2 = y2,
                bounds = {
                    x1 = x1,
                    y1 = y1,
                    x2 = x2,
                    y2 = y2,
                },
            }
        end

        local ax, ay = rotate_point(x1, y1, center_x, center_y, cos_a, sin_a)
        local bx, by = rotate_point(x2, y1, center_x, center_y, cos_a, sin_a)
        local cx, cy = rotate_point(x2, y2, center_x, center_y, cos_a, sin_a)
        local dx, dy = rotate_point(x1, y2, center_x, center_y, cos_a, sin_a)
        local min_x, min_y, max_x, max_y = rotated_rect_aabb(x1, y1, x2, y2, center_x, center_y, cos_a, sin_a)

        return {
            points = {
                { x = ax, y = ay },
                { x = bx, y = by },
                { x = cx, y = cy },
                { x = dx, y = dy },
            },
            bounds = {
                x1 = min_x,
                y1 = min_y,
                x2 = max_x,
                y2 = max_y,
            },
        }
    end

    if matrix and map and type(map) == "table" then
        local cell_count = cols * rows
        for cell = 0, cell_count - 1 do
            local mapped = map[cell + 1]
            if type(mapped) == "number" and mapped >= 0 then
                local led_idx = floor(mapped)
                if mapped == led_idx and led_idx < total_leds and not rects[led_idx + 1] then
                    local col = cell % cols
                    local row = floor(cell / cols)
                    local lx1 = origin_x + col * led_w
                    local ly1 = origin_y + row * led_h
                    local lx2 = origin_x + (col + 1) * led_w
                    local ly2 = origin_y + (row + 1) * led_h
                    rects[led_idx + 1] = make_led_geometry(lx1, ly1, lx2, ly2)
                end
            end
        end
    end

    -- Fallback for missing/unmapped LEDs to keep output length stable.
    for i = 0, total_leds - 1 do
        if not rects[i + 1] then
            local col = i % cols
            local row = floor(i / cols)
            local lx1 = origin_x + col * led_w
            local ly1 = origin_y + row * led_h
            local lx2 = origin_x + (col + 1) * led_w
            local ly2 = origin_y + (row + 1) * led_h
            rects[i + 1] = make_led_geometry(lx1, ly1, lx2, ly2)
        end
    end

    return rects
end

--- Build a routing table from the current canvas placements.
--
-- @param placements  Array of placement tables, each containing:
--   { port, outputId, segmentId?, x, y, width, height, rotation?, ledsCount, matrix,
--     localIndices?, actualIndices? }
-- @param grid_width  Canvas grid width (64 by default).
-- @param grid_height Canvas grid height (defaults to grid_width).
--
-- @return table  Keyed by `"port::outputId"` (or `"port::outputId::segmentId"`),
--   value = array of LED entries:
--     { overlaps = { { idx=canvas_flat_index, w=weight }, ... } }
--   `idx` is 0-based (row * grid_width + col), `w` is normalised ∈ (0,1].
function M.build(placements, grid_width, grid_height)
    grid_width = tonumber(grid_width) or 64
    grid_height = tonumber(grid_height) or grid_width
    grid_width = max(1, floor(grid_width + 0.5))
    grid_height = max(1, floor(grid_height + 0.5))
    local rt = {}

    for _, p in ipairs(placements) do
        local key = p.port .. "::" .. p.outputId
        if p.segmentId and p.segmentId ~= "" then
            key = key .. "::" .. p.segmentId
        end

        local cols, rows = led_grid_dims(p)
        local total_leds = p.ledsCount or (cols * rows)
        local led_rects = build_led_rects(p, cols, rows, total_leds)

        local actual_indices = type(p.actualIndices) == "table" and p.actualIndices or nil
        local local_indices = type(p.localIndices) == "table" and p.localIndices or nil
        if not local_indices then
            local_indices = {}
            for i = 0, total_leds - 1 do
                local_indices[#local_indices + 1] = i
            end
        end

        local actual_index_by_local = {}
        for idx, local_idx in ipairs(local_indices) do
            actual_index_by_local[local_idx] = actual_indices and actual_indices[idx] or local_idx
        end

        local entries = {}
        for _, local_idx in ipairs(local_indices) do
            local rect = led_rects[local_idx + 1]
            if not rect then
                goto continue
            end
            local bounds = rect.bounds or rect
            local lx1, ly1 = bounds.x1, bounds.y1
            local lx2, ly2 = bounds.x2, bounds.y2

            -- Canvas cell range that may overlap (cells are unit squares at integer coords)
            local cx_min = max(0, floor(lx1))
            local cy_min = max(0, floor(ly1))
            local cx_max = min(grid_width - 1, floor(lx2 - EPS))
            local cy_max = min(grid_height - 1, floor(ly2 - EPS))
            if cx_min > cx_max or cy_min > cy_max then
                goto continue
            end

            local overlaps = {}
            local total_w = 0

            for cy = cy_min, cy_max do
                for cx = cx_min, cx_max do
                    local area
                    if rect.points then
                        area = polygon_rect_overlap_area(rect.points, cx, cy, cx + 1, cy + 1)
                    else
                        area = overlap_area(rect.x1, rect.y1, rect.x2, rect.y2, cx, cy, cx + 1, cy + 1)
                    end
                    if area > EPS then
                        overlaps[#overlaps + 1] = { idx = cy * grid_width + cx, w = area }
                        total_w = total_w + area
                    end
                end
            end

            -- Normalise weights so they sum to 1.
            if total_w > 0 then
                local inv = 1 / total_w
                for _, o in ipairs(overlaps) do
                    o.w = o.w * inv
                end
            end

            if #overlaps <= 0 then
                goto continue
            end

            entries[#entries + 1] = {
                local_idx = local_idx,
                target_idx = actual_index_by_local[local_idx] or local_idx,
                overlaps = overlaps,
            }

            ::continue::
        end

        rt[key] = {
            port = p.port,
            outputId = p.outputId,
            segmentId = p.segmentId,
            placementId = p.id,
            localLedCount = p.ledsCount or total_leds,
            brightness = p.brightness or 100,
            leds = entries,
        }
    end

    return rt
end

--- Mix canvas frame colours through the routing table.
--
-- @param canvas_colors  Flat array of R,G,B bytes (1-based, length = grid_width * grid_height * 3).
-- @param rt             Routing table from `build()`.
-- @return table  Keyed identically to `rt`, values are arrays of
--   { target_idx, r, g, b } tuples ready for `ext.set_leds`.
function M.route(canvas_colors, rt)
    local result = {}

    for key, entry in pairs(rt) do
        local leds = entry.leds
        local n = #leds
        local brightness = entry.brightness or 100
        local factor = brightness < 100 and (brightness / 100) or nil
        local colors = {}
        local preview_colors = {}
        local preview_count = max(0, floor((entry.localLedCount or 0) + 0.5))

        for i = 1, preview_count do
            preview_colors[i] = { r = 0, g = 0, b = 0 }
        end

        for i = 1, n do
            local overlaps = leds[i].overlaps
            local r, g, b = 0, 0, 0

            for _, o in ipairs(overlaps) do
                local base = o.idx * 3 + 1    -- canvas_colors is 1-based, idx is 0-based
                local cr = canvas_colors[base]     or 0
                local cg = canvas_colors[base + 1] or 0
                local cb = canvas_colors[base + 2] or 0
                r = r + cr * o.w
                g = g + cg * o.w
                b = b + cb * o.w
            end

            if factor then
                r = r * factor
                g = g * factor
                b = b * factor
            end

            local rr = floor(r + 0.5)
            local gg = floor(g + 0.5)
            local bb = floor(b + 0.5)

            colors[#colors + 1] = {
                leds[i].target_idx or (i - 1),
                rr,
                gg,
                bb,
            }

            local preview_idx = (leds[i].local_idx or (i - 1)) + 1
            if preview_idx > 0 then
                preview_colors[preview_idx] = {
                    r = rr,
                    g = gg,
                    b = bb,
                }
            end
        end

        result[key] = {
            port = entry.port,
            outputId = entry.outputId,
            segmentId = entry.segmentId,
            placementId = entry.placementId,
            colors = colors,
            preview = preview_colors,
        }
    end

    return result
end

return M

File: src/routing.rs
use std::collections::HashMap;

use crate::abi::{SkydimoLedColorV1, SkydimoRgb};
use crate::model::{placement_routing_key, RuntimePlacement, Matrix};

const EPS: f64 = 1e-9;

#[derive(Clone, Debug, Default)]
pub struct RoutingTable {
    pub entries: HashMap<String, RouteEntry>,
}

#[derive(Clone, Debug)]
pub struct RouteEntry {
    pub port: String,
    pub output_id: String,
    pub placement_id: String,
    pub local_led_count: usize,
    pub brightness: u8,
    pub leds: Vec<RouteLed>,
}

#[derive(Clone, Debug)]
pub struct RouteLed {
    pub local_idx: usize,
    pub target_idx: usize,
    overlaps: Vec<Overlap>,
}

#[derive(Clone, Copy, Debug)]
struct Overlap {
    idx: usize,
    weight: f64,
}

#[derive(Clone, Debug)]
struct LedGeometry {
    rect: Rect,
    points: Option<[Point; 4]>,
}

#[derive(Clone, Copy, Debug)]
struct Rect {
    x1: f64,
    y1: f64,
    x2: f64,
    y2: f64,
}

#[derive(Clone, Copy, Debug)]
struct Point {
    x: f64,
    y: f64,
}

pub struct RouteOutput {
    pub port: String,
    pub output_id: String,
    pub placement_id: String,
    pub colors: Vec<SkydimoLedColorV1>,
    pub preview_rgb: String,
}

pub fn build(placements: &[RuntimePlacement], grid_width: usize, grid_height: usize) -> RoutingTable {
    let grid_width = grid_width.max(1);
    let grid_height = grid_height.max(1);
    let mut table = RoutingTable::default();

    for placement in placements {
        let key = placement_routing_key(placement);
        let (cols, rows) = led_grid_dims(placement);
        let total_leds = if placement.leds_count > 0 {
            placement.leds_count
        } else {
            cols.saturating_mul(rows)
        };
        let led_rects = build_led_rects(placement, cols, rows, total_leds);
        let actual_by_local = placement
            .local_indices
            .iter()
            .zip(placement.actual_indices.iter())
            .map(|(local, actual)| (*local, *actual))
            .collect::<HashMap<_, _>>();
        let mut leds = Vec::with_capacity(placement.local_indices.len());

        for local_idx in &placement.local_indices {
            let Some(rect) = led_rects.get(*local_idx).and_then(Option::as_ref) else {
                continue;
            };
            let bounds = rect.rect;
            let cx_min = bounds.x1.floor().max(0.0) as usize;
            let cy_min = bounds.y1.floor().max(0.0) as usize;
            let cx_max_float = (bounds.x2 - EPS).floor().min((grid_width - 1) as f64);
            let cy_max_float = (bounds.y2 - EPS).floor().min((grid_height - 1) as f64);
            if cx_max_float < cx_min as f64 || cy_max_float < cy_min as f64 {
                continue;
            }
            let cx_max = cx_max_float as usize;
            let cy_max = cy_max_float as usize;
            let mut overlaps = Vec::new();
            let mut total_weight = 0.0;

            for cy in cy_min..=cy_max {
                for cx in cx_min..=cx_max {
                    let area = if let Some(points) = rect.points {
                        polygon_rect_overlap_area(&points, cx as f64, cy as f64, cx as f64 + 1.0, cy as f64 + 1.0)
                    } else {
                        overlap_area(
                            rect.rect,
                            Rect {
                                x1: cx as f64,
                                y1: cy as f64,
                                x2: cx as f64 + 1.0,
                                y2: cy as f64 + 1.0,
                            },
                        )
                    };
                    if area > EPS {
                        overlaps.push(Overlap {
                            idx: cy.saturating_mul(grid_width).saturating_add(cx),
                            weight: area,
                        });
                        total_weight += area;
                    }
                }
            }

            if total_weight <= 0.0 {
                continue;
            }
            let inv = 1.0 / total_weight;
            for overlap in &mut overlaps {
                overlap.weight *= inv;
            }
            leds.push(RouteLed {
                local_idx: *local_idx,
                target_idx: actual_by_local.get(local_idx).copied().unwrap_or(*local_idx),
                overlaps,
            });
        }

        table.entries.insert(key, RouteEntry {
            port: placement.port.clone(),
            output_id: placement.output_id.clone(),
            placement_id: placement.id.clone(),
            local_led_count: placement.leds_count,
            brightness: placement.brightness,
            leds,
        });
    }

    table
}

pub fn route(canvas: &[SkydimoRgb], table: &RoutingTable) -> Vec<RouteOutput> {
    let mut outputs = Vec::with_capacity(table.entries.len());

    for entry in table.entries.values() {
        let mut color_values = Vec::with_capacity(entry.leds.len());
        let mut preview = vec![SkydimoRgb::default(); entry.local_led_count];
        let brightness = if entry.brightness < 100 {
            Some(f64::from(entry.brightness) / 100.0)
        } else {
            None
        };

        for led in &entry.leds {
            let mut r = 0.0;
            let mut g = 0.0;
            let mut b = 0.0;

            for overlap in &led.overlaps {
                if let Some(color) = canvas.get(overlap.idx) {
                    r += f64::from(color.r) * overlap.weight;
                    g += f64::from(color.g) * overlap.weight;
                    b += f64::from(color.b) * overlap.weight;
                }
            }

            if let Some(factor) = brightness {
                r *= factor;
                g *= factor;
                b *= factor;
            }

            let color = SkydimoRgb {
                r: to_u8(r),
                g: to_u8(g),
                b: to_u8(b),
            };
            color_values.push(SkydimoLedColorV1 {
                index: led.target_idx,
                color,
            });
            if let Some(slot) = preview.get_mut(led.local_idx) {
                *slot = color;
            }
        }

        outputs.push(RouteOutput {
            port: entry.port.clone(),
            output_id: entry.output_id.clone(),
            placement_id: entry.placement_id.clone(),
            colors: color_values,
            preview_rgb: rgb_vec_base64(&preview),
        });
    }

    outputs
}

pub fn rgb_vec_base64(colors: &[SkydimoRgb]) -> String {
    let mut encoded = String::with_capacity(colors.len().saturating_mul(4));
    for color in colors {
        let packed = (u32::from(color.r) << 16) | (u32::from(color.g) << 8) | u32::from(color.b);
        encoded.push(BASE64[(packed >> 18) as usize] as char);
        encoded.push(BASE64[((packed >> 12) & 0x3f) as usize] as char);
        encoded.push(BASE64[((packed >> 6) & 0x3f) as usize] as char);
        encoded.push(BASE64[(packed & 0x3f) as usize] as char);
    }
    encoded
}

const BASE64: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

fn led_grid_dims(placement: &RuntimePlacement) -> (usize, usize) {
    if let Some(matrix) = placement.matrix.as_ref() {
        if matrix.width > 0 && matrix.height > 0 {
            return (matrix.width, matrix.height);
        }
    }
    if placement.leds_count == 0 {
        (1, 1)
    } else {
        (placement.leds_count, 1)
    }
}

fn build_led_rects(
    placement: &RuntimePlacement,
    cols: usize,
    rows: usize,
    total_leds: usize,
) -> Vec<Option<LedGeometry>> {
    let dev_w = if placement.width > 0.0 { placement.width } else { cols as f64 };
    let dev_h = if placement.height > 0.0 { placement.height } else { rows as f64 };
    let origin_x = placement.x;
    let origin_y = placement.y;
    let rotation = placement.rotation.rem_euclid(360.0);
    let has_rotation = rotation.abs() > EPS;
    let angle = rotation.to_radians();
    let cos_a = angle.cos();
    let sin_a = angle.sin();
    let center_x = origin_x + dev_w / 2.0;
    let center_y = origin_y + dev_h / 2.0;
    let led_w = dev_w / cols.max(1) as f64;
    let led_h = dev_h / rows.max(1) as f64;
    let mut rects = vec![None; total_leds];

    if let Some(matrix) = placement.matrix.as_ref() {
        build_matrix_rects(
            matrix,
            &mut rects,
            total_leds,
            origin_x,
            origin_y,
            led_w,
            led_h,
            has_rotation,
            center_x,
            center_y,
            cos_a,
            sin_a,
        );
    }

    for (led_index, slot) in rects.iter_mut().enumerate().take(total_leds) {
        if slot.is_none() {
            let col = led_index % cols.max(1);
            let row = led_index / cols.max(1);
            *slot = Some(make_led_geometry(
                origin_x + col as f64 * led_w,
                origin_y + row as f64 * led_h,
                origin_x + (col + 1) as f64 * led_w,
                origin_y + (row + 1) as f64 * led_h,
                has_rotation,
                center_x,
                center_y,
                cos_a,
                sin_a,
            ));
        }
    }

    rects
}

#[allow(clippy::too_many_arguments)]
fn build_matrix_rects(
    matrix: &Matrix,
    rects: &mut [Option<LedGeometry>],
    total_leds: usize,
    origin_x: f64,
    origin_y: f64,
    led_w: f64,
    led_h: f64,
    has_rotation: bool,
    center_x: f64,
    center_y: f64,
    cos_a: f64,
    sin_a: f64,
) {
    let cell_count = matrix.width.saturating_mul(matrix.height);
    for cell in 0..cell_count.min(matrix.map.len()) {
        let mapped = matrix.map[cell];
        if mapped < 0 {
            continue;
        }
        let Ok(led_index) = usize::try_from(mapped) else {
            continue;
        };
        if led_index >= total_leds || rects.get(led_index).and_then(Option::as_ref).is_some() {
            continue;
        }
        let col = cell % matrix.width.max(1);
        let row = cell / matrix.width.max(1);
        rects[led_index] = Some(make_led_geometry(
            origin_x + col as f64 * led_w,
            origin_y + row as f64 * led_h,
            origin_x + (col + 1) as f64 * led_w,
            origin_y + (row + 1) as f64 * led_h,
            has_rotation,
            center_x,
            center_y,
            cos_a,
            sin_a,
        ));
    }
}

#[allow(clippy::too_many_arguments)]
fn make_led_geometry(
    x1: f64,
    y1: f64,
    x2: f64,
    y2: f64,
    has_rotation: bool,
    center_x: f64,
    center_y: f64,
    cos_a: f64,
    sin_a: f64,
) -> LedGeometry {
    if !has_rotation {
        return LedGeometry {
            rect: Rect { x1, y1, x2, y2 },
            points: None,
        };
    }

    let points = [
        rotate_point(x1, y1, center_x, center_y, cos_a, sin_a),
        rotate_point(x2, y1, center_x, center_y, cos_a, sin_a),
        rotate_point(x2, y2, center_x, center_y, cos_a, sin_a),
        rotate_point(x1, y2, center_x, center_y, cos_a, sin_a),
    ];
    let rect = points.iter().fold(
        Rect {
            x1: f64::INFINITY,
            y1: f64::INFINITY,
            x2: f64::NEG_INFINITY,
            y2: f64::NEG_INFINITY,
        },
        |mut acc, point| {
            acc.x1 = acc.x1.min(point.x);
            acc.y1 = acc.y1.min(point.y);
            acc.x2 = acc.x2.max(point.x);
            acc.y2 = acc.y2.max(point.y);
            acc
        },
    );

    LedGeometry {
        rect,
        points: Some(points),
    }
}

fn rotate_point(px: f64, py: f64, cx: f64, cy: f64, cos_a: f64, sin_a: f64) -> Point {
    let dx = px - cx;
    let dy = py - cy;
    Point {
        x: cx + dx * cos_a - dy * sin_a,
        y: cy + dx * sin_a + dy * cos_a,
    }
}

fn overlap_area(a: Rect, b: Rect) -> f64 {
    let ox = 0.0_f64.max(a.x2.min(b.x2) - a.x1.max(b.x1));
    let oy = 0.0_f64.max(a.y2.min(b.y2) - a.y1.max(b.y1));
    ox * oy
}

fn polygon_rect_overlap_area(points: &[Point; 4], x1: f64, y1: f64, x2: f64, y2: f64) -> f64 {
    let mut clipped = clip_polygon(points.as_slice(), |point| point.x >= x1 - EPS, |a, b| {
        let dx = b.x - a.x;
        let t = (x1 - a.x) / dx;
        Point { x: x1, y: a.y + (b.y - a.y) * t }
    });
    if clipped.is_empty() {
        return 0.0;
    }
    clipped = clip_polygon(&clipped, |point| point.x <= x2 + EPS, |a, b| {
        let dx = b.x - a.x;
        let t = (x2 - a.x) / dx;
        Point { x: x2, y: a.y + (b.y - a.y) * t }
    });
    if clipped.is_empty() {
        return 0.0;
    }
    clipped = clip_polygon(&clipped, |point| point.y >= y1 - EPS, |a, b| {
        let dy = b.y - a.y;
        let t = (y1 - a.y) / dy;
        Point { x: a.x + (b.x - a.x) * t, y: y1 }
    });
    if clipped.is_empty() {
        return 0.0;
    }
    clipped = clip_polygon(&clipped, |point| point.y <= y2 + EPS, |a, b| {
        let dy = b.y - a.y;
        let t = (y2 - a.y) / dy;
        Point { x: a.x + (b.x - a.x) * t, y: y2 }
    });
    polygon_area(&clipped)
}

fn clip_polygon(
    points: &[Point],
    inside: impl Fn(Point) -> bool,
    intersection: impl Fn(Point, Point) -> Point,
) -> Vec<Point> {
    let Some(mut prev) = points.last().copied() else {
        return Vec::new();
    };
    let mut result = Vec::with_capacity(points.len() + 2);
    let mut prev_inside = inside(prev);

    for curr in points.iter().copied() {
        let curr_inside = inside(curr);
        if prev_inside != curr_inside {
            result.push(intersection(prev, curr));
        }
        if curr_inside {
            result.push(curr);
        }
        prev = curr;
        prev_inside = curr_inside;
    }

    result
}

fn polygon_area(points: &[Point]) -> f64 {
    if points.len() < 3 {
        return 0.0;
    }
    let mut area = 0.0;
    let mut prev = points[points.len() - 1];
    for curr in points {
        area += prev.x * curr.y - curr.x * prev.y;
        prev = *curr;
    }
    area.abs() * 0.5
}

fn to_u8(value: f64) -> u8 {
    (value + 0.5).floor().clamp(0.0, 255.0) as u8
}
